docs/architecture/adr-versions.md

```diff
@@ -0,0 +1,219 @@
+	- Tendermint should be stabilizing - it's just Atomic Broadcast.
+	- We can start considering for Tendermint v2.0 in a year
+- For now, assume that BlockVersion will not change in the life of a given chain, except by critical emergency.
+- To effectively change it, dump state and start a new blockchain with the new BlockVersion.
+- It's easy to determine the version of a block from its serialized form
+
+#### P2PVersion
+
+- All peer messaging (messages, detectable behaviour)
+- Will change gradually as reactors evolve to improve performance and support new features
+	- eg proposed new message types BatchTx in the mempool and HasBlockPart in the consensus
+- It's easy to determine the version of a peer from its first serialized message/s
+- New versions must be compatible with at least one old version to allow gradual upgrades
+
+#### AppVersion
+
+- The ABCI state machine (txs, begin/endblock behaviour, commit hashing)
+- Behaviour and message types will change abruptly in the course of the life of a chain
+- Need to minimize complexity of the code for supporting different AppVersions at different heights
+- Ideally, each version of the software supports only a *single* AppVersion at one time
+    - this means we checkout different versions of the software at different heights instead of littering the code
+          with conditionals
+    - minimize the number of data migrations required across AppVersion (ie. most AppVersion should be able to read the same state from disk as previous AppVersion).
+- NOTE: we still need to support peers syncing from old heights across AppVersion changes
+
+### Networks
+
+We need to support many independent networks running the same version of the software,
+even possibly starting from the same initial state.
+They must have distinct identifiers so that peers know which one they are joining and so
+validators and users can prevent replay attacks.
+
+Call this the `NetworkName`. It represents both the application being run and the community or intention
+of running it.
+
+Peers only connect to other peers with the same NetworkName.
+
+### Forks
+
+We need to support existing networks upgrading and forking, wherein they may do any of:
+	- revert back to some height, continue with the same versions but new blocks
+	- arbitrarily mutate state at some height, continue with the same versions (eg. Dao Fork)
+	- change the AppVersion at some height
+
+Note because of Tendermint's voting power threshold rules, a chain can only be extended under the "original" rules and under the new rules
+if 1/3 or more is double signing, which is expressly prohibited, and is supposed to result in their punishment on both chains. Since they can censor
+the punishment, the chain is expected to be hardforked to remove the validators. Thus, if both branches are to continue after a fork,
+they will each require a new identifier, and the old chain identifier will be retired (ie. only useful for syncing history, not for new blocks)..
```

Comment:
I'm not sure this is what will actually happen in a "contentious fork", either or both "sides" may elect to have any validators who sign on the other chain be slashable on their chain.

Of course, "contentious" forks may disregard versioning rules anyways.

Reply:
It is not really clear how will slashing work in case the forks adopt new chain identifiers? Since the new blocks belong to different chains so two blocks at a given height `h` belonging to different chains signed by the same validator set cannot be considered as a reason to slash? I would be a great help if a little more detail could be added to this.

Reply:
great point. chains will have to remember their "original" chain id so they can punish. I'll add a note